cmd/root.go

```diff
@@ -50,96 +53,209 @@ const (
 	waitRemoteLoggerTimeout = time.Second * 5
 )
 
-// TODO better name - there are other command flags these are just ... non lib.Options ones :shrug:
-type commandFlags struct {
-	defaultConfigFilePath string
-	configFilePath        string
-	exitOnRunning         bool
-	showCloudLogs         bool
-	runType               string
-	archiveOut            string
-	quiet                 bool
-	noColor               bool
-	address               string
-	outMutex              *sync.Mutex
-	stdoutTTY, stderrTTY  bool
-	stdout, stderr        *consoleWriter
+// globalFlags contains global config values that apply for all k6 sub-commands.
+type globalFlags struct {
+	configFilePath string
+	runType        string
+	quiet          bool
+	noColor        bool
+	address        string
+	logOutput      string
+	logFormat      string
+	verbose        bool
 }
 
-func newCommandFlags() *commandFlags {
-	confDir, err := os.UserConfigDir()
-	if err != nil {
-		logrus.WithError(err).Warn("could not get config directory")
-		confDir = ".config"
-	}
-	defaultConfigFilePath := filepath.Join(
-		confDir,
-		"loadimpact",
-		"k6",
-		defaultConfigFileName,
-	)
+// globalState contains the globalFlags and accessors for most of the global
+// process-external state like CLI arguments, env vars, standard input, output
+// and error, etc. In practice, most of it is normally accessed through the `os`
+// package from the Go stdlib.
+//
+// We group them here so we can prevent direct access to them from the rest of
+// the k6 codebase. This gives us the ability to mock them and have robust and
+// easy-to-write integration-like tests to check the k6 end-to-end behavior in
+// any simulated conditions.
+//
+// `newGlobalState()` returns a globalState object with the real `os`
+// parameters, while `newGlobalTestState()` can be used in tests to create
+// simulated environments.
+type globalState struct {
+	ctx context.Context
+
+	fs      afero.Fs
+	getwd   func() (string, error)
+	args    []string
+	envVars map[string]string
+
+	defaultFlags, flags globalFlags
+
+	outMutex       *sync.Mutex
+	stdOut, stdErr *consoleWriter
+	stdIn          io.Reader
+
+	signalNotify func(chan<- os.Signal, ...os.Signal)
+	signalStop   func(chan<- os.Signal)
+
+	// TODO: add os.Exit()?
 
+	logger         *logrus.Logger
+	fallbackLogger logrus.FieldLogger
+}
+
+// Ideally, this should be the only function in the whole codebase where we use
+// global variables and functions from the os package. Anywhere else, things
+// like os.Stdout, os.Stderr, os.Stdin, os.Getenv(), etc. should be removed and
+// the respective properties of globalState used instead.
+func newGlobalState(ctx context.Context) *globalState {
```

Comment:
Any particular reason we don't want to just provide `envs map[string]string` to this so we can drop `os.Get...` here as well? I guess we will still need access to os.Stdout and co , but this might be worth abstracting either way :shrug:

Reply:
I could, but I kind of like that this is the only place in the code that touches the `os` package... passing the env vars here won't really have any effect besides a slightly cleaner internals. If you dislike using `os.Getenv` in `DisableColors`, I'd rather move up the this line higher in the function than pass it from outside;
```go
envVars := buildEnvMap(os.Environ())
```
but I don't think it matters :man_shrugging:

Reply:
I guess we can do this later if needed. 

Moving the remaining `os` things will in practice move most of the code from this function to w/e calls it, likely not improving anything, but the opposite